reference/programming-model.md

```diff
@@ -33,80 +64,243 @@ The [@pulumi/pulumi] package is the core library for working with the Pulumi pla
+- `dependsOn` - a list of explicit resource dependencies
 - `protect` - whether to mark a resource as protected. A protected resource cannot be deleted directly: first you must set `protect: false` and run `pulumi update`. Then, the resource can be deleted, either by removing the line of code or by running `pulumi destroy`.
 - `parent` - optional parent for the resource. See [Components](#components).
 - `provider` - optional provider for the resource. See [Providers](#providers).
 
 ## Resource outputs {#outputs}
 
-<!-- TODO: add direct anchor link to `apply` once #333 is fixed -->
+Outputs are a key part of how Pulumi tracks dependencies between resources.  Because the values of Outputs are not available until resources are created, these are represented using a special [Output] type which internally represents two things:
+1. An eventually available value of the output
+2. The depdency on the source(s) of the output value
+
+The output properties of all resource objects in Pulumi have type [`Output`][pulumi.Output]. Resource inputs have type [`Input`][pulumi.Input], which accepts either a raw value or an output from another resource. This allows dependencies to be inferred, including ensuring that resources are not created or updatred until all their dependencies are availabe and up to date.  
+
+To transform an output into a new value, use the [`apply` method](pkg/nodejs/@pulumi/pulumi/#property-apply). For example, use the following to create an HTTPS URL from the DNS name of a virtual machine: 
+
+{% include langchoose.html %}
+
+```javascript
+let url = virtualmachine.dnsName.apply(dnsName => "https://" + dnsName)
+```
+
+```typescript
+let url = virtualmachine.dnsName.apply(dnsName => "https://" + dnsName)
+```
+
+```python
+url = virtual_machine.dns_name.apply(
+    lambda dns_name: "https://" + dns_name
+)
+```
+
+```go
+url := virtualmachine.DnsName().Apply(func(dnsName string) (interface{}, error) { 
+    return "https://" + dnsName, nil 
+})
+```
+
+The `apply` method accepts a callback which will be passed the value of the `Output` when it is availabe, and which returns the new value.  The result of the call to `apply` is new `Output` whose value is the value returned from the callback, and which includes the dependencies of the original `Output`.  If the callback itself returns an `Output`, the dependencies of that output are unioned into the dependencies of the returned `Output`.
+
+> Note: The `Output` itself cannot be used directly in string concatenation or other operations, as it is not itself the value of the output.  To transform the value of the output (when it becomes available), the `apply` method should be used instead.
+
+To combine multiple `Output`s into a transformed value, use [pulumi.all].  This allows a new value to be constructed from several inputs, such as concatenating outputs from two different resources together, or constructing a policy document using information from several other resources.
```

Comment:
I wonder if `apply` and `all` should get their own H3 headers, so we can more easily link to them.